tests: integration: build_test: add the kw build test

aquilamacedo opened the pull request:
This PR Introduce integration tests for the kw build feature with randomized
distribution selection. The tests select a random distribution from Debian,
Archlinux, and Fedora, optimizing resource usage by running tests on a single
selected distro.

aquilamacedo committed:
tests: integration: build_test: add the kw build test

Introduce integration tests for the kw build feature with randomized
distribution selection. The tests select a random distribution from
Debian, Archlinux, and Fedora, optimizing resource usage by running
tests on a single selected distro.

Signed-off-by: Aquila Macedo <aquilamacedo@riseup.net>

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  container_exec "$container" 'yes | ./setup.sh --install-kernel-dev-deps'
+  if [[ "$?" -ne 0 ]]; then
+    complain "Failed to install kernel build dependencies for ${distro}"
+    return 1 # EPERM
```
EINVAL sounds better here.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+    return 1 # EPERM
+  fi
+
+  eval "git clone --depth 1 ${url_kernel_repo_tree} ${cloned_kernel_tree_path_host}"
```
We have a build feature named `--from-sha`, and to test it, we might use a large depth, maybe increasing it to 5.

rodrigosiqueira replied:
Why do you need eval here? Why not use git directly?

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  eval "git clone --depth 1 ${url_kernel_repo_tree} ${cloned_kernel_tree_path_host}"
+  if [[ "$?" -ne 0 ]]; then
+    complain "Failed to clone ${url_kernel_repo_tree}"
+    rm -rf "${cloned_kernel_tree_path_host}"
```
Never call rm with a bash variable before checking the variable. Also, use is_safe_path_to_remove after checking the variable, and finally remove the target.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  fi
+}
+
+function prepare_environment()
```
This name is too generic; use a more descriptive name, maybe something like `setup_kernel_tree_with_config_file`.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  local build_status_log
+
+  container="kw-${target_random_distro}"
+  container_tmp_dir=$(container_exec "$container" 'mktemp --directory')
```
This variable name could be more descriptive. Find a better name for it.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+
+  kw_build_cmd='kw build --cpu-scaling 50'
+  container_exec "$container" "cd ${container_tmp_dir}/linux && ${kw_build_cmd}"
+  assert_equals_helper "kw build failed failed for ${container}" "($LINENO)" 0 "$?"
```
Add `{}` around LINENO (check this in all of your changes).

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  build_status_log=$(container_exec "$container" "sqlite3 ~/.local/share/kw/kw.db \"SELECT * FROM statistics_report\" | tail --lines=1")
+
+  # Extract the build status and result from the log
+  build_status=$(printf '%s' "$build_status_log" | cut -d'|' -f2)
```
https://kworkflow.org/content/codingstyle.html#avoid-short-options

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+
+  # Extract the build status and result from the log
+  build_status=$(printf '%s' "$build_status_log" | cut -d'|' -f2)
+  assert_equals_helper "Build status check failed for $container" "$LINENO" 'build' "$build_status"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  build_result=$(printf '%s' "$build_status_log" | cut -d'|' -f3)
+  assert_equals_helper "Build result check failed for $container" "$LINENO" 'success' "$build_result"
+
+  # Clean the kernel tree to the next test
+  container_exec "$container" "cd ${container_tmp_dir}/linux && kw build --clean"
+  assert_equals_helper "kw build ---clean failed failed for ${container}" "($LINENO)" 0 "$?"
```
This type of operation should be in a tearDown function.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  assert_equals_helper "kw build --clean failed for ${container}" "($LINENO)" 0 "$?"
+}
+
+function test_kernel_build_only_cpu_scaling_option()
```
I was thinking about this test, and you should find a way to identify CPU utilization when compiling the kernel. If it is around half of the CPU capacity, it should pass. You don't even need to run the entire test for that.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+oneTimeTearDown()
+{
+  say "Cleaning..."
+  rm -rf "${cloned_kernel_tree_path_host}"
```
check my previous comments about rm

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  assert_equals_helper "kw build ---clean failed failed for ${container}" "($LINENO)" 0 "$?"
+}
+
+function test_kernel_build_cpu_scaling_llvm()
```
Check my previous comments, they apply to this test.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,189 @@
+  local build_result
+  local build_status_log
+
+  # Define the container name using the target random distribution
+  container="kw-${target_random_distro}"
+  # Create a temporary directory inside the container
+  container_tmp_dir=$(container_exec "$container" 'mktemp --directory')
```
I think this should be in a setUp function.

rodrigosiqueira commented on tests/integration/build_test.sh:
Overall, you should use more of the setUp/tearDown functions. Also, review the codestyle for all your changes in this file.

rodrigosiqueira commented on documentation/dependencies/kernel_build/arch.dependencies:
This commit should come first.

aquilamacedo committed:
documentation: dependencies: update kernel build and integration test dependencies

This commit updates the documentation to include additional dependencies
required for kernel building and integration tests for kw build

Signed-off-by: Aquila Macedo <aquilamacedo@riseup.net>

aquilamacedo committed:
tests: integration: podman: update containerfiles to handle test scripts

Updated Containerfile_archlinux, Containerfile_debian, and
Containerfile_fedora to copy and manage integration test scripts,
ensuring they are executable and correctly placed in /bin.

Signed-off-by: Aquila Macedo <aquilamacedo@riseup.net>

rodrigosiqueira commented on tests/integration/podman/Containerfile_archlinux:
```diff
@@ -5,3 +5,11 @@ RUN pacman -Syu --noconfirm git
 
 RUN bash ./clone_and_install_kw.sh
+
+COPY ./scripts /scripts
```
I'm confused about this part; where does this `scripts` folder come from? Is this the script folder from kw that has `kwreview.sh  profiler.sh`? If so, why do you need it?

aquilamacedo replied:
No, this `/scripts` is from the path `tests/integration/podman/scripts`

rodrigosiqueira commented on tests/integration/podman/Containerfile_archlinux:
```diff
@@ -5,3 +5,11 @@ RUN pacman -Syu --noconfirm git
+
+COPY ./scripts /scripts
+
+COPY ./move_test_scripts_to_bin.sh /tmp/move_test_scripts_to_bin.sh
```
Are you sure you want to keep this script in the `/tmp` folder from the container? From the script name, it sounds like you want to put it in /usr/bin or similar.

rodrigosiqueira commented on tests/integration/podman/Containerfile_archlinux:
```diff
@@ -5,3 +5,11 @@ RUN pacman -Syu --noconfirm git
+
+RUN bash ./tmp/move_test_scripts_to_bin.sh
+
+RUN rm -rf /scripts
```
Avoid short options.

rodrigosiqueira commented on tests/integration/podman/move_test_scripts_to_bin.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+# This script moves all shell script files (.sh) from the source directory to
```
I think it would be easier if you just removed the .sh extension from all files from the script folder directly and kept them executable instead of running post-processing as you did here.

rodrigosiqueira commented on tests/integration/podman/move_test_scripts_to_bin.sh:
```diff
@@ -0,0 +1,18 @@
+SRC_DIR='/scripts'
+DEST_DIR='/bin'
+
+for file in "$SRC_DIR"/*.sh; do
+  base_name=$(basename "$file" .sh)
+  mv "$file" "${DEST_DIR}/${base_name}"
+  chmod +x "${DEST_DIR}/${base_name}"
+done
```
tbh, I don't understand why we need a script for this... you could follow my previous comment and just use the copy commands in the container.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,175 @@
+  setup_container_environment "$VERBOSE" 'build' "$TARGET_RANDOM_DISTRO"
+
+  # Install kernel build dependencies
+  container_exec "$CONTAINER" 'yes | ./setup.sh --install-kernel-dev-deps'
```
Can we hide this command's output and only expose it if the user uses `-v`?

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,175 @@
+    complain "Failed to clone ${url_kernel_repo_tree}"
+    if [[ -n "$CLONED_KERNEL_TREE_PATH_HOST" ]]; then
+      if is_safe_path_to_remove "$CLONED_KERNEL_TREE_PATH_HOST"; then
+        rm -rf "$CLONED_KERNEL_TREE_PATH_HOST"
```
avoid short options.

rodrigosiqueira commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,175 @@
+  fi
+}
+
+function setup_kernel_tree_with_config_file()
```
I think you should call this function inside the `setUp` and remove it from the specific tests.

aquilamacedo commented on documentation/dependencies/fedora.dependencies:
```diff
@@ -27,3 +27,4 @@ b4
 procps-ng
 pciutils
 libnotify
+psmisc
```
The `psmisc` dependency is added because it includes `pstree`, which is needed to check the process tree.

aquilamacedo replied:
For some reason, this dep is only missing on Fedora. My theory is that when installing multiple dependencies on other distros (like **Arch** and **Debian**), some package likely pulls in `psmisc` package as a dependency, so it gets installed automatically. However, this doesn’t happen on **Fedora**, so we need to add the dependency manually

aquilamacedo commented:
Hi @rodrigosiqueira and @davidbtadokoro.

I made the changes, could you check?

rodrigosiqueira commented:
Hi @aquilamacedo ,
I have this error on my side:

![Screenshot from 2024-07-20 09-03-16](https://github.com/user-attachments/assets/7da99026-4e88-41fa-9b75-7559aa0b9104)


@davidbtadokoro can you check if you can repro this issue on your side?

aquilamacedo commented on tests/integration/podman/scripts/kw_build_cpu_scaling_monitor:
```diff
@@ -0,0 +1,113 @@
+declare -g CPU_USAGE_FILE='/tmp/cpu_usage.txt'
+declare -g KW_BUILD_CMD="kw build --cpu-scaling ${CPU_SCALING}"
+
+# CPU Usage (%) = (Δ CPU Time / (Δ Time × Number of CPUs)) × 100
```
that's really helpful :-)

rodrigosiqueira commented on tests/integration/podman/scripts/kw_build_cpu_scaling_monitor:
```diff
@@ -0,0 +1,117 @@
+
+if [[ "$#" -eq 0 ]]; then
+  printf 'Usage: %s <cpu_scaling_value>\n' "$0"
+  exit 1
```
22 sounds better

rodrigosiqueira commented on tests/integration/podman/scripts/kw_build_cpu_scaling_monitor:
```diff
@@ -0,0 +1,117 @@
+      IFS=' ' read -r -a pids_to_kill <<< "$(pstree --show-pids "$parent_pid" | grep --only-matching '([0-9]\+)' | grep --only-matching '[0-9]\+')"
+
+      printf "Cancelling PIDs: %s\n" "${pids_to_kill[@]}"
+      printf "%s\n" "${pids_to_kill[@]}" | xargs kill -9
```
Replace all `"` with `'` for all the `printf`s in this entire commit.

rodrigosiqueira reviewed (changes requested):
Hi @aquilamacedo ,

I added some comments.

aquilamacedo commented on tests/integration/podman/scripts/kw_build_cpu_scaling_monitor:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
```
Like, a dedicated file to explain how this test works, or just add comments at the beginning of the code?

aquilamacedo commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,183 @@
+  fi
+
+  # Optimize the configuration based on the currently loaded modules
+  container_exec "$CONTAINER" "cd ${KERNEL_TREE_PATH_CONTAINER} && make LSMOD=${kernel_test_tmp_dir}/container_mod_list localmodconfig > /dev/null 2>&1"
```
I updated the variable name earlier but forgot to reflect the change here. Thx for catching that!

aquilamacedo commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,183 @@
+  assert_equals_helper "Build status check failed for ${CONTAINER}" "$LINENO" 'build' "$build_status"
+
+  build_result=$(printf '%s' "$build_status_log" | cut --delimiter='|' --fields=3)
+  assert_equals_helper "Build result check failed for ${CONTAINER}" "$LINENO" 'success' "$build_result"
```
I considered modularizing this part, but I realized it would bring back the issue where `assert_equals_helper` always points to the same line when an error occurs, making debugging significantly more challenging.

aquilamacedo commented on tests/integration/build_test.sh:
```diff
@@ -0,0 +1,183 @@
+  assert_equals_helper "Build result check failed for ${CONTAINER}" "$LINENO" 'success' "$build_result"
+}
+
+function test_kernel_build_cpu_scaling_option()
```
i changed for `test_kw_build_cpu_scaling_execution` what do you think?

aquilamacedo commented on tests/integration/podman/Containerfile_archlinux:
```diff
@@ -5,3 +5,12 @@ RUN pacman -Syu --noconfirm git
+COPY scripts/ /tmp/scripts/
+
+# Grant execution permissions to the copied scripts
+RUN chmod +x /tmp/scripts/*
```
Actually, I don’t think they are executable. When I don’t set the permission, We get a `Permission denied` error.

![image](https://github.com/user-attachments/assets/ffa284cb-d5c2-4f46-bbd9-63e35ef1d2d2)

rodrigosiqueira commented:
Hi,
Nobody has reviewed this change for a long time, and I have also not heard anything about this PR author. I'll close it for now. If you want to reopen the PR, we can resume the conversation in the future.

rodrigosiqueira commented:
Hi @aquilamacedo ,
I have this issue on my side:

```
Running test [integration/build_test]
=========================================================
egrep: warning: egrep is obsolescent; using grep -E
test_kernel_build_gcc_x86_64_no_env
test_kw_build_cpu_scaling_execution
test_kernel_build_llvm
podman container exec kw-debian /bin/sh -c $'cd /tmp/tmp.Y2QCCm3oZN/linux && kw build --llvm > /dev/null 2>&1' 2> /dev/null
ASSERT:(426): Failed to execute the command in the container.
ASSERT: line (163): kw build failed for kw-debian
  Expected Result: 0
    Actual Result: 1
ASSERT: line 165: Kernel binary not found in /tmp/tmp.Y2QCCm3oZN/linux/arch/x86/boot/ for kw-debian
  Expected Result: 0
    Actual Result: 1
ASSERT: line 165: Build result from database failed for kw-debian
  Expected Result: success
    Actual Result: failure
```

Any ideas?

rodrigosiqueira commented:
Hi,
This PR has been open for a long time. I'm going to close it for now. Maybe we'll bring it back at some point.
